File: display/app/src/main/java/com/driveon/FpsTestView.java
package com.driveon;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.view.View;

import java.util.Random;

public class FpsTestView extends View {

    private static final int BALL_COUNT = 50;

    private final float[] x = new float[BALL_COUNT];
    private final float[] y = new float[BALL_COUNT];
    private final float[] vx = new float[BALL_COUNT];
    private final float[] vy = new float[BALL_COUNT];

    private final Paint ballPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint textPaint = new Paint(Paint.ANTI_ALIAS_FLAG);

    private final Random random = new Random();

    private boolean running = false;

    // FPS
    private long lastFpsTime = 0;
    private int frames = 0;
    private int fps = 0;

    // Loop
    private long lastFrameTime = 0;
    private static final long FRAME_DELAY_MS = 16; // ~60fps

    public FpsTestView(Context context) {
        super(context);

        ballPaint.setColor(Color.RED);

        textPaint.setColor(Color.WHITE);
        textPaint.setTextSize(32f);

        for (int i = 0; i < BALL_COUNT; i++) {
            x[i] = random.nextInt(800);
            y[i] = random.nextInt(480);
            vx[i] = random.nextFloat() * 8f - 4f;
            vy[i] = random.nextFloat() * 8f - 4f;
        }
    }

    public void start() {
        running = true;
        lastFrameTime = System.currentTimeMillis();
        lastFpsTime = lastFrameTime;
        invalidate();
    }

    public void stop() {
        running = false;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        if (!running) return;

        canvas.drawColor(Color.BLACK);

        long now = System.currentTimeMillis();
        update(now - lastFrameTime);
        lastFrameTime = now;

        drawBalls(canvas);
        drawFps(canvas);

        frames++;
        if (now - lastFpsTime >= 1000) {
            fps = frames;
            frames = 0;
            lastFpsTime = now;
        }

        postInvalidateDelayed(FRAME_DELAY_MS);
    }

    private void update(long deltaMs) {
        int w = getWidth();
        int h = getHeight();

        for (int i = 0; i < BALL_COUNT; i++) {
            x[i] += vx[i];
            y[i] += vy[i];

            if (x[i] < 0 || x[i] > w) vx[i] = -vx[i];
            if (y[i] < 0 || y[i] > h) vy[i] = -vy[i];
        }
    }

    private void drawBalls(Canvas canvas) {
        for (int i = 0; i < BALL_COUNT; i++) {
            canvas.drawCircle(x[i], y[i], 10f, ballPaint);
        }
    }

    private void drawFps(Canvas canvas) {
        canvas.drawText("FPS: " + fps, 20, 40, textPaint);
        canvas.drawText("Bolas: " + BALL_COUNT, 20, 80, textPaint);
    }
}

File: display/app/src/main/java/com/example/driveondisplay/FpsTestView.kt
package com.example.driveondisplay

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.os.SystemClock
import android.view.MotionEvent
import android.view.View
import kotlin.math.abs
import kotlin.random.Random

class FpsTestView(context: Context) : View(context) {

    private data class Ball(
        var x: Float,
        var y: Float,
        var vx: Float,
        var vy: Float,
        val r: Float
    )

    // ===== CONFIG =====
    private val BALL_COUNT = 40
    private val BG_COLOR = Color.rgb(15, 15, 15)

    // ===== PAINTS =====
    private val paintBall = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.RED
    }

    private val paintText = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.WHITE
        textSize = 42f
    }

    // ===== STATE =====
    private val balls = ArrayList<Ball>(BALL_COUNT)

    private var lastFrameTime = 0L
    private var fps = 0
    private var frames = 0
    private var lastFpsTime = 0L

    init {
        repeat(BALL_COUNT) {
            balls.add(
                Ball(
                    x = Random.nextFloat() * 800f,
                    y = Random.nextFloat() * 600f,
                    vx = Random.nextFloat() * 12f - 6f,
                    vy = Random.nextFloat() * 12f - 6f,
                    r = Random.nextFloat() * 12f + 8f
                )
            )
        }
    }

    override fun onDraw(canvas: Canvas) {
        val now = SystemClock.elapsedRealtime()

        if (lastFrameTime == 0L) {
            lastFrameTime = now
            lastFpsTime = now
        }

        val dt = (now - lastFrameTime) / 1000f
        lastFrameTime = now

        canvas.drawColor(BG_COLOR)

        updateBalls(dt)
        drawBalls(canvas)
        calculateFps(now)
        drawFps(canvas)

        // ✅ render sincronizado com VSYNC
        postInvalidateOnAnimation()
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        parent?.requestDisallowInterceptTouchEvent(true)
        return true
    }

    private fun updateBalls(dt: Float) {
        val w = width.toFloat()
        val h = height.toFloat()

        for (b in balls) {
            b.x += b.vx * dt * 60f
            b.y += b.vy * dt * 60f

            if (b.x - b.r < 0 || b.x + b.r > w) b.vx = -b.vx
            if (b.y - b.r < 0 || b.y + b.r > h) b.vy = -b.vy
        }
    }

    private fun drawBalls(c: Canvas) {
        for (b in balls) {
            c.drawCircle(b.x, b.y, b.r, paintBall)
        }
    }

    private fun calculateFps(now: Long) {
        frames++
        if (now - lastFpsTime >= 1000) {
            fps = frames
            frames = 0
            lastFpsTime = now
        }
    }

    private fun drawFps(c: Canvas) {
        c.drawText("FPS: $fps", 30f, 60f, paintText)
        c.drawText("Bolas: $BALL_COUNT", 30f, 110f, paintText)
    }
}
